### ISSUE TITLE
Unions of Callables reporting error in contravariant generic

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#3844]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from collections.abc import Callable
from typing import Any, Generic, TypeVar
from typing_extensions import reveal_type

T_contra = TypeVar('T_contra', contravariant=True)


class Thing1:
    prop1: str
    prop2: str


class Thing2:
    prop1: str
    prop2: str
    prop3: str


class CallbackManager(Generic[T_contra]):
    def __init__(self, callback: Callable[[T_contra], Any]) -> None:
        self._callback: Callable[[T_contra], Any] = callback


def main(
    cb1: Callable[[Thing1], Any],
    cb2: Callable[[Thing1 | Thing2], Any],
    cb3: Callable[[Thing1], Any] | Callable[[Thing1 | Thing2], Any],
) -> None:
    reveal_type(CallbackManager(cb1))  # CallbackManager[Thing1]
    reveal_type(CallbackManager(cb2))  # CallbackManager[Thing1 | Thing2]
    reveal_type(CallbackManager(cb3))  # CallbackManager[Thing1] in 1.1.266, error in 1.1.267
```

Line 33 reports the following error:

```
error: Argument of type "((Thing1) -> Any) | ((Thing1 | Thing2) -> Any)" cannot be assigned to parameter "callback" of type "(T_contra@CallbackManager) -> Any" in function "__init__"
    Type "((Thing1) -> Any) | ((Thing1 | Thing2) -> Any)" cannot be assigned to type "(T_contra@CallbackManager) -> Any"
      Type "(Thing1) -> Any" cannot be assigned to type "(T_contra@CallbackManager) -> Any"
        Parameter 1: type "T_contra@CallbackManager" cannot be assigned to type "Thing1"
          Type "Thing1 | Thing2" cannot be assigned to type "Thing1"
            "Thing2" is incompatible with "Thing1"
```

This seems to be a regression from 1.1.266 where no error is reported and the revealed type is `CallbackManager[Thing1]`.

**To Reproduce**
Steps to reproduce the behavior.

**Expected behavior**
No error should be reported and the revealed type should be `CallbackManager[Thing1]`

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.267

**Comment by erictraut** [comment]
Prior to 1.1.267, the behavior depended on the ordering of callables within a union. It worked with one order and not with the other. In 1.1.267, I eliminated this non-determinism. You will now consistently receive an error regardless of ordering within the union.

That said, I think you're correct that this should type check without error. I'll investigate further.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.269, which I just published. It will also be included in a future version of pylance.

**Comment by bryanforbes** [comment]
Thank you! And thanks for all the work you're doing on this project.
